File: 1-easy/problems/0020-valid-parentheses/solution.ts
function isValid(s: string): boolean {
    const stack: string[] = [];
    const pairs: {[k:string]:string} = {')':'(', ']':'[', '}':'{'};
    for (const c of s) {
        if ('([{'.includes(c)) stack.push(c);
        else if (!stack.length || stack[stack.length-1] !== pairs[c]) return false;
        else stack.pop();
    }
    return stack.length === 0;
}

File: 1-easy/problems/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {')':'(', ']':'[', '}':'{'}
        for c in s:
            if c in '([{': stack.append(c)
            elif not stack or stack[-1] != pairs[c]: return False
            else: stack.pop()
        return not stack
